### ISSUE TITLE
Add support for vertical writing

### ISSUE THREAD

**Issue by Rufy86** [lingochamp/FuriganaTextView#6]
Thank you very very very much for this textview. Finally there is a view that in the correct way put the furigana on Kanji.

Thank you.

I'm trying to use this text view but if try to set scrollEnabled = false. but it's not work. or if I try to change the backgroundColor, the color don't change.

how can I do these things?

an other thing: I red that unfotunally 'Furiganas displayed in vertical writing is not implemented yet.' when will it be available?

**Comment by eyeplum** [comment]
Hi @Rufy86,

I'm glad this library is useful for you. Thank you.

As for the customizations (scroll, background color, etc.) of the text view, I'd like to first talk about some details about how `FuriganaTextView` is implemented.

`FuriganaTextView` is merely a wrapper view of a real `UITextView`. If you look at the [source code](https://github.com/lingochamp/FuriganaTextView/blob/master/src/FuriganaTextView.swift#L19), you will find that the `FuriganaTextView` class itself is not really a `UITextView`, it's actually a `UIView`. Basically, we just put a `UITextView` inside a `UIView` and added some extra logics to handle furiganas in the `UIView`, thus it becomes the `FuriganaTextView`.

With this design, in order to customize the appearance of the final result, we need to change the `UITextView` instead of the wrapper `UIView`. This is the reason why `backgroundColor` of the `FuriganaTextView` doesn't work (technically, it changed the background color of the wrapper view, but it's invisible).

We did noticed this problem when we were implementing this view, but the customizations we needed was limited to our use cases by then, so we came up with some wrapper properties like `scrollEnabled` and `alignment` to get things done.

This issue you reported gets me rethink about this problem. And I realize that a full functioning reference to the underlying `UITextView` is a way better solution than some wrapper properties.

I have made [some changes](https://github.com/lingochamp/FuriganaTextView/pull/7):

-  A new read-only property is added: `contentView: UITextView?`, it will return a valid `UITextView` for you to customize **after** the `content` has been set.
- The following wrapper properties are deprecated.
    - `scrollEnabled`
    - `alignment`

To try it out, please use the latest version [0.4.1](https://github.com/lingochamp/FuriganaTextView/releases/tag/0.4.1).

I also updated the [sample code](https://github.com/lingochamp/FuriganaTextView/blob/master/demo/demo/ViewController.swift#L25) to make use of the new `contentView` property.

Hope it can fit your needs. Thanks!

**Comment by eyeplum** [comment]
As for the vertical writing. I'm planning to look into this soon. But I'm afraid I don't have a timeframe for the release yet.

**Comment by Rufy86** [comment]
It’s a shame. Because the vertical writing is a feature very important for my project. Unfotunally don’t know Text Kit but I can learn. Then, can you give me some indication about to do it? So I can try to do the changes to insert this feature. Obviously, I’ll send you the code, if I’ll found the solution.

What do you think about it?

**Comment by eyeplum** [comment]
That would be great.

I suggest to start with layout the original text vertically (without any furiganas) at first. As far as I know, `NSTextContainer` (which implements `NSTextLayoutOrientationProvider`) would be a good start to tweak with.

Please don't hesitate to follow up here if you want to discuss anything.

And when you have something completely implemented, please open a pull request so we can review it and eventually merge it into the repo.

Thank you so much for your help.

**Comment by Rufy86** [comment]
I'm sorry for late. I was busy. I've made some search about the vertical text and I've found this:
I've found this answer [here:](http://stackoverflow.com/a/28694927)

TextKit way

The proper way to keep all native text benefits (e.g. highlighting, selection...) is to use standard TextKit APIs. The method you are proposing would break all that or would possibly result in strange behaviour.

However, looks like TextKit in iOS does not support vertical orientation out-of-the-box yet, but it is prepared for that. As a side note, in OS X it is somewhat supported and you could call textView.setLayoutOrientation(.Vertical), but it still has some limitations.

The NSTextLayoutOrientationProvider protocol defines an interface providing the default orientation for text laid out in a conforming object, in absence of an explicit NSVerticalGlyphFormAttributeName attribute. The only UIKit class that implements this interface is  NSTextContainer, whose default implementation returns NSTextLayoutOrientationHorizontal. An NSTextContainer subclass that handles vertical text could set this property to  NSTextLayoutOrientationVertical to support the custom layout orientation logic.
Source: UIKit > NSTextLayoutOrientationProvider Protocol Reference for iOS

In conclusion, you should start subclassing NSTextContainer, and you will have to deal with NSLayoutManager and NSTextContainer a lot.

unfortunately I've no knowledge about text kit, but I'll try to do something.

**Comment by 2small2fall** [comment]
Hello folks,

Until one succeeds in implementing a pure Textkit solution, here is an elegant vertical attributed text alternative one :

The trick relies on "NSVerticalGlyphFormAttributeName" attribute being set to true when it is not supposed to be supported by IOS (false by default and result off setting value to "1/true" : undefined) but OSX. https://developer.apple.com/reference/appkit/nsverticalglyphformattributename

(pure CoreText way : https://developer.apple.com/reference/coretext/kctverticalformsattributename?language=objc. To be preferred I guess, toll-free bridging : simply replace nsverticalglyphformattributename by kctverticalformsattributename)

However, the following code (ViewController.swift) works perfectly on Xcode 8.2 >= IOS 8.0 (successfully tested on an iPhone6 running >= iOS 8.0).

(SOURCE : https://teratail.com/questions/62049)

Happy coding and thank you VERY MUCH for your GREAT FuriganaTextView project. 

NOTE : I haven't able to find a way to use the NSStrikethroughStyleAttributeName attribute vertically.

// ____________________________________________________________________________
import UIKit

extension String {
    func find(pattern: String) -> NSTextCheckingResult? {
        do {
            let re = try NSRegularExpression(pattern: pattern, options: [])
            return re.firstMatch(
                in: self,
                options: [],
                range: NSMakeRange(0, self.utf16.count))
        } catch {
            return nil
        }
    }

    func replace(pattern: String, template: String) -> String {
        do {
            let re = try NSRegularExpression(pattern: pattern, options: [])
            return re.stringByReplacingMatches(
                in: self,
                options: [],
                range: NSMakeRange(0, self.utf16.count),
                withTemplate: template)
        } catch {
            return self
        }
    }
}

class View: UIView {

    override func draw(_ rect: CGRect) {
        let text = [
            "「まさか、｜後罪《クライム》の｜触媒《カタリスト》を〈｜讃来歌《オラトリオ》〉無しで？」",
            "教師たちの狼狽した声が次々と上がる。",
            "……なんでだろう。何を驚いているんだろう。",
            "ただ普通に、この｜触媒《カタリスト》を使って｜名詠門《チャネル》を開かせただけなのに。",
            "そう言えば、何を｜詠《よ》ぼう。",
            "自分の一番好きな花でいいかな。",
            "どんな宝石より素敵な、わたしの大好きな緋色の花。",
            "――『｜Keinez《赤》』――",
            "そして、少女の口ずさんだその後に――",
            ]
            .joined(separator: "\n")

        let attributed =
            text
                .replace(pattern: "(｜.+?《.+?》)", template: ",$1,")
                .components(separatedBy: ",")
                .map { x -> NSAttributedString in
                    if let pair = x.find(pattern: "｜(.+?)《(.+?)》") {
                        let string = (x as NSString).substring(with: pair.rangeAt(1))
                        let ruby = (x as NSString).substring(with: pair.rangeAt(2))

                        var text: [Unmanaged<CFString>?] = [Unmanaged<CFString>.passRetained(ruby as CFString) as Unmanaged<CFString>, .none, .none, .none]

                        let annotation = CTRubyAnnotationCreate(.auto, .auto, 0.5, &text[0]!)

                        return NSAttributedString(
                            string: string,
                            attributes: [kCTRubyAnnotationAttributeName as String: annotation])
                    } else {
                        return NSAttributedString(string: x, attributes: nil)
                    }
                }
                .reduce(NSMutableAttributedString()) { $0.append($1); return $0 }

        var height = 28.0
        let settings = [
            CTParagraphStyleSetting(
                spec: .minimumLineHeight,
                valueSize: Int(MemoryLayout.size(ofValue: height)),
                value: &height)
        ]
        let style = CTParagraphStyleCreate(settings, Int(settings.count))

        attributed.addAttributes([
            NSFontAttributeName: UIFont(name: "HiraMinProN-W3", size: 14.0)!,
            NSVerticalGlyphFormAttributeName: true,
            kCTParagraphStyleAttributeName as String: style,
            ],
                                 range: NSMakeRange(0, attributed.length))

        let context = UIGraphicsGetCurrentContext()

        context!.setFillColor(red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0)
        context!.addRect(rect)
        context?.fillPath()

        context!.rotate(by: CGFloat(M_PI_2))
        context!.translateBy(x: 30.0, y: 35.0)
        context!.scaleBy(x: 1.0, y: -1.0)

        let framesetter = CTFramesetterCreateWithAttributedString(attributed)
        let path = CGPath(rect: CGRect(x: 0.0, y: 0.0, width: rect.height, height: rect.width), transform: nil)
        let frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, 0), path, nil)
        CTFrameDraw(frame, context!)
    }
}

class ViewController: UIViewController {
    override func loadView() {
        super.loadView()

        self.view = View()
    }
}
// ____________________________________________________________________________

**Comment by Rufy86** [comment]
@eyeplum  if you can implement this correctly, you can create a great component. I'm trying to use it but I've some problem
this is my code.

as you see, I've added an enum to set the vertical or horizontal method.

```
import UIKit
extension String {
    func find(pattern: String) -> NSTextCheckingResult? {
        do {
            let re = try NSRegularExpression(pattern: pattern, options: [])
            return re.firstMatch(
                in: self,
                options: [],
                range: NSMakeRange(0, self.utf16.count))
        } catch {
            return nil
        }
    }
    
    func replace(pattern: String, template: String) -> String {
        do {
            let re = try NSRegularExpression(pattern: pattern, options: [])
            return re.stringByReplacingMatches(
                in: self,
                options: [],
                range: NSMakeRange(0, self.utf16.count),
                withTemplate: template)
        } catch {
            return self
        }
    }
}

enum textOrientation:Int {
    case horizontal
    case vertical
}

class GQAsianTextLabel: UILabel {
    var orientation:textOrientation = .horizontal
    
    @IBInspectable var orientationAdapter:Int{
        get{
            return self.orientation.rawValue
        }
        set (orientationIndex){
            self.orientation = textOrientation(rawValue: orientationIndex) ?? .horizontal
        }
    }
    
    
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        let attributed =
            self.text?
                .replace(pattern: "(｜.+?《.+?》)", template: ",$1,")
                .components(separatedBy: ",")
                .map { x -> NSAttributedString in
                    if let pair = x.find(pattern: "｜(.+?)《(.+?)》") {
                        let string = (x as NSString).substring(with: pair.rangeAt(1))
                        let ruby = (x as NSString).substring(with: pair.rangeAt(2))
                        
                        var text: [Unmanaged<CFString>?] = [Unmanaged<CFString>.passRetained(ruby as CFString) as Unmanaged<CFString>, .none, .none, .none]
                        
                        let annotation = CTRubyAnnotationCreate(CTRubyAlignment.auto, CTRubyOverhang.auto, 0.5, &text[0]!)
                        
                        return NSAttributedString(
                            string: string,
                            attributes: [kCTRubyAnnotationAttributeName as String: annotation])
                    } else {
                        return NSAttributedString(string: x, attributes: nil)
                    }
                }
                .reduce(NSMutableAttributedString()) { $0.append($1); return $0 }
        
        var height = self.font.lineHeight
        var heightMultiple = 3.0
        
        var lineSpacingAdjustment = 15.0
        
        let horizontalSettings = [CTParagraphStyleSetting(spec: .maximumLineHeight, valueSize: Int(MemoryLayout.size(ofValue: height)), value: &height),
                                  /*CTParagraphStyleSetting(spec: .lineHeightMultiple, valueSize: Int(MemoryLayout.size(ofValue: heightMultiple)), value: &heightMultiple)*/
        ]
        let horizontalStyle = CTParagraphStyleCreate(horizontalSettings, Int(horizontalSettings.count))
        
        let verticalSettings = [CTParagraphStyleSetting(spec: .maximumLineHeight, valueSize: Int(MemoryLayout.size(ofValue: height)), value: &height),
                                CTParagraphStyleSetting(spec: .minimumLineSpacing, valueSize: Int(MemoryLayout.size(ofValue: lineSpacingAdjustment)), value: &lineSpacingAdjustment)
        ]
        let verticalStyle = CTParagraphStyleCreate(verticalSettings, Int(verticalSettings.count))
        
        
        
        
        
        
        let settings = [
            CTParagraphStyleSetting(spec: .maximumLineHeight, valueSize: Int(MemoryLayout.size(ofValue: height)), value: &height),
            CTParagraphStyleSetting(spec: .minimumLineSpacing, valueSize: Int(MemoryLayout.size(ofValue: lineSpacingAdjustment)), value: &lineSpacingAdjustment)
            //CTParagraphStyleSetting(spec: .minimumLineSpacing, valueSize: <#T##Int#>, value: <#T##UnsafeRawPointer#>)
        ]
        let style = CTParagraphStyleCreate(settings, Int(settings.count))
        
        
        self.numberOfLines = 0
        
        switch orientation {
        case .horizontal:
            attributed?.addAttributes([
                NSFontAttributeName: self.font,
                NSVerticalGlyphFormAttributeName: false,
                kCTParagraphStyleAttributeName as String: horizontalStyle,
                ],
                                      range: NSMakeRange(0, (attributed?.length)!))
        case .vertical:
            attributed?.addAttributes([
                NSFontAttributeName: self.font,
                NSVerticalGlyphFormAttributeName: true,
                kCTParagraphStyleAttributeName as String: verticalStyle,
                ],
                                      range: NSMakeRange(0, (attributed?.length)!))
        }
        
        
        // Initialize a graphics context in iOS.
        let context = UIGraphicsGetCurrentContext()
        
        let bounds:CGRect
        
        // Set the text matrix.
        context!.textMatrix = CGAffineTransform.identity;
        
        // Flip the context coordinates, in iOS only.
        switch orientation {
        case .horizontal:
            context!.translateBy(x: 0, y: 0);
            context!.scaleBy(x: 1.0, y: -1.0);
            //bounds = CGRect(x: 0, y: 0, width: rect.size.width, height: -rect.size.height)
            bounds = CGRect(x: 0, y: 0, width: rect.size.width, height: -CGFloat((context?.height)!))
            
            
        case .vertical:
            context!.rotate(by: CGFloat(Double.pi / 2))
            context!.translateBy(x: 0, y: 0);
            context!.scaleBy(x: 1.0, y: -1.0);
            bounds = CGRect(x: 0, y: 0, width: rect.size.height, height: rect.size.width)
            
        }
        
        let framesetter = CTFramesetterCreateWithAttributedString(attributed!)
        
        let path = CGPath(rect: bounds, transform: nil)
        let frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, 0), path, nil)
        CTFrameDraw(frame, context!)
        
        
        
        
        
        
        
        
        // Create a path which bounds the area where you will be drawing text.
        // The path need not be rectangular.
        //let path = CGMutablePath()
        
        // In this simple example, initialize a rectangular path.
        //let bounds = CGRect(x: rect.origin.x, y: rect.origin.y, width: rect.size.width, height: -(rect.size.height * 2))
        //path.addRect(bounds)
        
        // Create the framesetter with the attributed string.
        //let framesetter = CTFramesetterCreateWithAttributedString(attributed!)
        
        
        // Create a frame.
        //let frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, 0), path, nil)
        // Draw the specified frame in the given context.
        //CTFrameDraw(frame, context!);
    }
}
```

but I've some problem. example:
![screen shot 2017-04-04 at 13 28 14](https://cloud.githubusercontent.com/assets/17876640/24729805/32725a34-1a60-11e7-89bd-2b8427e95866.png)
![screen shot 2017-04-06 at 00 15 20](https://cloud.githubusercontent.com/assets/17876640/24729806/32b5a906-1a60-11e7-8d9b-29fea865168f.png)


can you fix the problem?

**Comment by eyeplum** [comment]
Thanks for your effort on this @Rufy86.
Could you fork this repo and push your changes to your fork?
I will take a look and see what I can do.
Thanks!

**Comment by Rufy86** [comment]
@eyeplum you are welcome! for me this feature is very very important. and I need it in few time. this is the reason of my effort.

**Comment by eyeplum** [comment]
Hey @Rufy86, hope everything is good.

I'm at WWDC17 Text and Fonts Lab now, I've just asked some Apple Engineer about the `NSTextLayoutOrientationVertical`. The response is that this option is not supported on iOS because the `TextKit` implementation is based on the assumption that the layout will be horizontal.

So, I guess the only possible solution will be to convert the `FuriganaTextView` to an implementation based on `CoreText` and a custom `UIView`. Since the initial idea about this library is only about presenting (not editing), that shouldn't be too hard. Especially we have `CTRubyAnnotation` to support the Furigana part.

I will start working on this when I have time. But if you have a tight deadline for this, I won't be able to help.

Thanks!

**Comment by eyeplum** [comment]
Closing this issue as it's been inactive for several years.
